### PR TITLE
fix: 修复SQL防火墙之SQL条件总真(WhereAlwayTrue)功能无效问题(条件写反了)

### PR DESCRIPTION
**Pull request by qxo**
包含这三项配置对应功能:deleteWhereAlwayTrue/updateWhereAlwayTrue/selectWhereAlwayTrue 并修正了PGWallVisitor/SQLServerWallVisitor

### CODE REVIEW

**Comment by zrlw** [comment]
30个单元测试都失败，确定单元测试都写错了？

**Comment by qxo** [comment]
> 30个单元测试都失败，确定单元测试都写错了？

此事我很是诧异, 好像从没正确过....
已经把测试用例和修复代码提交了, 此事请项目负责人接管核实吧

**Comment by qxo** [comment]
> > 30个单元测试都失败，确定单元测试都写错了？
> 
> 此事我很是诧异, 好像从没正确过.... 已经把测试用例和修复代码提交了, 此事请项目负责人接管核实吧

应该是部分逻辑写反了, 而之前的测试用没覆盖到场景, 所以修正代码后所有测试用例也是正常的

**Comment by zrlw** [comment]
2024-01-20T05:58:34.1417283Z [ERROR] Failures: 
2024-01-20T05:58:34.1417981Z [ERROR]   TenantSelectTest4.testMySql:53 expected:<0> but was:<1>
2024-01-20T05:58:34.1418798Z [ERROR]   TenantSelectTest4.testMySql2:62 expected:<0> but was:<1>
2024-01-20T05:58:34.1419497Z [ERROR]   WallDeleteWhereTest.testMySql:32
2024-01-20T05:58:34.1420185Z [ERROR]   WallDeleteWhereTest.testORACLE:37
2024-01-20T05:58:34.1420933Z [ERROR]   WallDeleteWhereTest2.test_check_true:38
2024-01-20T05:58:34.1421624Z [ERROR]   WallSelectWhereTest.testMySql:33
2024-01-20T05:58:34.1422146Z [ERROR]   WallSelectWhereTest.testORACLE:37
2024-01-20T05:58:34.1422561Z [ERROR]   WallSelectWhereTest6.test_check_true:39
2024-01-20T05:58:34.1422954Z [ERROR]   WallUnionTest.testMySql:34
2024-01-20T05:58:34.1423292Z [ERROR]   WallUnionTest.testOracle:41
2024-01-20T05:58:34.1423634Z [ERROR]   WallUnionTest2.testMySql:34
2024-01-20T05:58:34.1424192Z [ERROR]   WallUnionTest2.testOracle:43
2024-01-20T05:58:34.1424644Z [ERROR]   WallUnionTest4.testMySql:34
2024-01-20T05:58:34.1424995Z [ERROR]   WallUnionTest4.testORACLE:43
2024-01-20T05:58:34.1425343Z [ERROR]   WallUpdateTest3.testMySql:41
2024-01-20T05:58:34.1425687Z [ERROR]   WallUpdateTest3.testORACLE:46
2024-01-20T05:58:34.1426063Z [ERROR]   WallUpdateWhereTest.testMySql:34
2024-01-20T05:58:34.1426457Z [ERROR]   WallUpdateWhereTest.testORACLE:39
2024-01-20T05:58:34.1426867Z [ERROR]   WallUpdateWhereTest1.test_check_true:38
2024-01-20T05:58:34.1427243Z [ERROR]   MySqlWallTest63.test_true:40
2024-01-20T05:58:34.1427675Z [ERROR]   MySqlWallTest74.test_false:38
2024-01-20T05:58:34.1428021Z [ERROR]   MySqlWallTest90.test_true:38
2024-01-20T05:58:34.1428347Z [ERROR]   MySqlWallTest91.test:245
2024-01-20T05:58:34.1428770Z [ERROR]   MySqlWallTest91.test_false3:78
2024-01-20T05:58:34.1429119Z [ERROR]   MySqlWallTest91.test_false4:86
2024-01-20T05:58:34.1429470Z [ERROR]   MySqlWallTest91.test_false5:94
2024-01-20T05:58:34.1429823Z [ERROR]   MySqlWallTest91.test_false6:102
2024-01-20T05:58:34.1430193Z [ERROR]   MySqlWallTest_union.testUnion:30
2024-01-20T05:58:34.1430557Z [ERROR]   OracleWallTest.testWall:29
2024-01-20T05:58:34.1430928Z [ERROR]   AntlrMySqlTest.test_for_antlr_examples:39

修订的单元测试没有提交？还是30个错误

**Comment by qxo** [comment]
github 都构建成功了, 应该提交了, 今天github无法访问, 没法确认



------------------&nbsp;原始邮件&nbsp;------------------
发件人: "Rossen ***@***.***&gt;; 
发送时间: 2024年1月24日(星期三) 中午11:52
收件人: ***@***.***&gt;; 
抄送: ***@***.***&gt;; ***@***.***&gt;; 
主题: Re: [alibaba/druid] fix: 修复SQL防火墙之SQL条件总真(WhereAlwayTrue)功能无效问题(条件写反了) (PR #5698)





 
2024-01-20T05:58:34.1417283Z [ERROR] Failures:
 2024-01-20T05:58:34.1417981Z [ERROR]   TenantSelectTest4.testMySql:53 expected:<0&gt; but was:<1&gt;
 2024-01-20T05:58:34.1418798Z [ERROR]   TenantSelectTest4.testMySql2:62 expected:<0&gt; but was:<1&gt;
 2024-01-20T05:58:34.1419497Z [ERROR]   WallDeleteWhereTest.testMySql:32
 2024-01-20T05:58:34.1420185Z [ERROR]   WallDeleteWhereTest.testORACLE:37
 2024-01-20T05:58:34.1420933Z [ERROR]   WallDeleteWhereTest2.test_check_true:38
 2024-01-20T05:58:34.1421624Z [ERROR]   WallSelectWhereTest.testMySql:33
 2024-01-20T05:58:34.1422146Z [ERROR]   WallSelectWhereTest.testORACLE:37
 2024-01-20T05:58:34.1422561Z [ERROR]   WallSelectWhereTest6.test_check_true:39
 2024-01-20T05:58:34.1422954Z [ERROR]   WallUnionTest.testMySql:34
 2024-01-20T05:58:34.1423292Z [ERROR]   WallUnionTest.testOracle:41
 2024-01-20T05:58:34.1423634Z [ERROR]   WallUnionTest2.testMySql:34
 2024-01-20T05:58:34.1424192Z [ERROR]   WallUnionTest2.testOracle:43
 2024-01-20T05:58:34.1424644Z [ERROR]   WallUnionTest4.testMySql:34
 2024-01-20T05:58:34.1424995Z [ERROR]   WallUnionTest4.testORACLE:43
 2024-01-20T05:58:34.1425343Z [ERROR]   WallUpdateTest3.testMySql:41
 2024-01-20T05:58:34.1425687Z [ERROR]   WallUpdateTest3.testORACLE:46
 2024-01-20T05:58:34.1426063Z [ERROR]   WallUpdateWhereTest.testMySql:34
 2024-01-20T05:58:34.1426457Z [ERROR]   WallUpdateWhereTest.testORACLE:39
 2024-01-20T05:58:34.1426867Z [ERROR]   WallUpdateWhereTest1.test_check_true:38
 2024-01-20T05:58:34.1427243Z [ERROR]   MySqlWallTest63.test_true:40
 2024-01-20T05:58:34.1427675Z [ERROR]   MySqlWallTest74.test_false:38
 2024-01-20T05:58:34.1428021Z [ERROR]   MySqlWallTest90.test_true:38
 2024-01-20T05:58:34.1428347Z [ERROR]   MySqlWallTest91.test:245
 2024-01-20T05:58:34.1428770Z [ERROR]   MySqlWallTest91.test_false3:78
 2024-01-20T05:58:34.1429119Z [ERROR]   MySqlWallTest91.test_false4:86
 2024-01-20T05:58:34.1429470Z [ERROR]   MySqlWallTest91.test_false5:94
 2024-01-20T05:58:34.1429823Z [ERROR]   MySqlWallTest91.test_false6:102
 2024-01-20T05:58:34.1430193Z [ERROR]   MySqlWallTest_union.testUnion:30
 2024-01-20T05:58:34.1430557Z [ERROR]   OracleWallTest.testWall:29
 2024-01-20T05:58:34.1430928Z [ERROR]   AntlrMySqlTest.test_for_antlr_examples:39
 
修订的单元测试没有提交？还是30个错误

**Comment by zrlw** [comment]
github显示你最后一次提交的是：
[qxo](https://github.com/qxo) [force-pushed](https://github.com/alibaba/druid/compare/9d624b4a510bb4fb1c136af2c9bde48b5d1375be..6393f022c4b4ec64a586d4da5408bd707aed778c) the fix-wall-deleteWhereAlwaysTrue branch from [9d624b4 ](https://github.com/alibaba/druid/commit/9d624b4a510bb4fb1c136af2c9bde48b5d1375be)to [6393f02 ](https://github.com/alibaba/druid/commit/6393f022c4b4ec64a586d4da5408bd707aed778c)
[5 days ago](https://github.com/alibaba/druid/pull/5698#event-11544986701)

github构建失败：
@github-actions
Java CI / Test JDK 17, ubuntu-latest (pull_request) Failing after 3m
点开查看detail有30个单元测试失败

github通常要fan才能正常访问。

**Comment by zrlw** [comment]
github页面显示只修改了5个文件，应该是你单元测试相关修订的push操作失败了，可以查看一下你本地的git log，应该和github显示的最后一个提交 [9d624b4 ](https://github.com/alibaba/druid/commit/9d624b4a510bb4fb1c136af2c9bde48b5d1375be) 不同。

**Comment by qxo** [comment]
> github页面显示只修改了5个文件，应该是你单元测试相关修订的push操作失败了，可以查看一下你本地的git log，应该和github显示的最后一个提交 [9d624b4 ](https://github.com/alibaba/druid/commit/9d624b4a510bb4fb1c136af2c9bde48b5d1375be) 不同。

我这显示和上游是一致的, 是5文件没错,但commitId为:6393f022c
为什么JDK17下我抽时间排查一下吧

**Comment by qxo** [comment]
我们先重新理此问题定义,&nbsp;WhereAlwayTrue=true时这个SQL算正常不: `where 1=1 limit 1,10`





------------------&nbsp;原始邮件&nbsp;------------------
发件人: "Rossen ***@***.***&gt;; 
发送时间: 2024年1月25日(星期四) 中午1:10
收件人: ***@***.***&gt;; 
抄送: ***@***.***&gt;; ***@***.***&gt;; 
主题: Re: [alibaba/druid] fix: 修复SQL防火墙之SQL条件总真(WhereAlwayTrue)功能无效问题(条件写反了) (PR #5698)





 
github页面显示只修改了5个文件，应该是你单元测试相关修订的push操作失败了，可以查看一下你本地的git log，应该和github显示的最后一个提交 9d624b4  不同。

**Comment by zrlw** [comment]
为了兼顾druid-spring-boot-3-starter，druid的github构建脚本只配了jdk17执行单元测试，jdk8环境下只编译代码，并不执行单元测试案例。

**Comment by zrlw** [comment]
当selectWhereAlwayTrueCheck、deleteWhereAlwayTrueCheck、updateWhereAlayTrueCheck（这个属性名称丢了一个w字母，最好加上）为true：检查相关where条件是否恒为真，此时相关where 1=1应该视为错误。
如果上述check选项为false，则不检查where条件，此时where 1=1应该视为正常。

**Comment by zrlw** [comment]
如果是[6393f02](https://github.com/alibaba/druid/commit/6393f022c4b4ec64a586d4da5408bd707aed778c)，那你本地跑mvn test也会有30个单元测试错误。

**Comment by zrlw** [comment]
这个PR需要修改的单元测试类都不只5个。

File: core/src/main/java/com/alibaba/druid/wall/WallProvider.java [hunk 1]
```diff
@@ -498,7 +498,7 @@ private WallCheckResult checkInternal(String sql) {
                 violations.add(new IllegalSQLObjectViolation(ErrorCode.SYNTAX_ERROR, "not terminal sql, token "
                         + lastToken, sql));
             }
-            endOfComment = parser.getLexer().isEndOfComment();
+            endOfComment = lastToken == Token.EOF || parser.getLexer().isEndOfComment();
```
Comment (zrlw) [inline core/src/main/java/com/alibaba/druid/wall/WallProvider.java:501]: 建议和其他endOfComment的判定条件保持一致：
```
endOfComment = parser.getLexer().isEOF(); 
```

**Review by zrlw — CHANGES_REQUESTED** [review]

File: core/src/main/java/com/alibaba/druid/wall/spi/WallVisitorUtils.java [hunk 2]
```diff
@@ -221,7 +221,7 @@ public static void checkSelelct(WallVisitor visitor, SQLSelectQueryBlock x) {
             if (Boolean.TRUE.equals(whereValue)) {
                 if (visitor.getConfig().isSelectWhereAlwayTrueCheck()
```
Comment (zrlw) [inline core/src/main/java/com/alibaba/druid/wall/spi/WallVisitorUtils.java:222]: visitor.getConfig().isSelectWhereAlwayTrueCheck() 前面getConditionValue也有作为入参，但并没有用到，建议这个PR同时把getConditionValue方法也一起改了，用不到的入参直接删掉吧。

**Review by zrlw — CHANGES_REQUESTED** [review]

**Comment by zrlw** [comment]
还有这个updateWhereAlayTrueCheck丢了字母的变量名，最好也一起改了

**Comment by qxo** [comment]
> 还有这个updateWhereAlayTrueCheck丢了字母的变量名，最好也一起改了
这是另一个问题,其实我们内部分支已做兼容性修正, 外部的话得看是否有业余时间另外提交对应PR

File: core/src/main/java/com/alibaba/druid/wall/WallProvider.java (hunk 1 above)
Comment (qxo) [reply core/src/main/java/com/alibaba/druid/wall/WallProvider.java:501]: 此处不调整我还不清楚如何兼容修改正问题, 有空时我再看看

**Comment by qxo** [comment]
> 当selectWhereAlwayTrueCheck、deleteWhereAlwayTrueCheck、updateWhereAlayTrueCheck（这个属性名称丢了一个w字母，最好加上）为true：检查相关where条件是否恒为真，此时相关where 1=1应该视为错误。 如果上述check选项为false，则不检查where条件，此时where 1=1应该视为正常。

如果这样的相关测试用例也写反了... 昨我还一度以为之前我对定义理解错了...以为: `where 1=1` 应该等同于没有where的场景

File: core/src/main/java/com/alibaba/druid/wall/WallProvider.java (hunk 1 above)
Comment (zrlw) [reply core/src/main/java/com/alibaba/druid/wall/WallProvider.java:501]: > 此处不调整我还不清楚如何兼容修改正问题, 有空时我再看看

druid其他代码都是用parser.getLexer().isEOF()方法设置endOfComment标识，此处写作parser.getLexer().isEndOfComment()应该是笔误。

**Comment by zrlw** [comment]
> > 当selectWhereAlwayTrueCheck、deleteWhereAlwayTrueCheck、updateWhereAlayTrueCheck（这个属性名称丢了一个w字母，最好加上）为true：检查相关where条件是否恒为真，此时相关where 1=1应该视为错误。 如果上述check选项为false，则不检查where条件，此时where 1=1应该视为正常。
> 
> 如果这样的相关测试用例也写反了... 昨我还一度以为之前我对定义理解错了...以为: `where 1=1` 应该等同于没有where的场景

不知道从什么版本开始这个检查功能就是个摆设了，侧面也说明这个功能很少有人用。

**Comment by zrlw** [comment]
[fix: 修正测试用户逻辑](https://github.com/alibaba/druid/pull/5698/commits/84707ed2936d1a2715b28275fe3f009acd9059a1)这个commit有20个测试ERROR：
2024-01-27T16:08:50.6612964Z [ERROR]   WallUnionTest2.testMySql:34
2024-01-27T16:08:50.6613602Z [ERROR]   WallUnionTest2.testOracle:43
2024-01-27T16:08:50.6614202Z [ERROR]   WallUnionTest4.testMySql:34
2024-01-27T16:08:50.6616489Z [ERROR]   WallUnionTest4.testORACLE:43
2024-01-27T16:08:50.6617366Z [ERROR]   WallUpdateTest3.testMySql:41
2024-01-27T16:08:50.6617998Z [ERROR]   WallUpdateTest3.testORACLE:46
2024-01-27T16:08:50.6618641Z [ERROR]   WallUpdateWhereTest.testMySql:34
2024-01-27T16:08:50.6619301Z [ERROR]   WallUpdateWhereTest.testORACLE:39
2024-01-27T16:08:50.6619990Z [ERROR]   WallUpdateWhereTest1.test_check_true:38
2024-01-27T16:08:50.6620651Z [ERROR]   MySqlWallTest63.test_true:40
2024-01-27T16:08:50.6621231Z [ERROR]   MySqlWallTest74.test_false:38
2024-01-27T16:08:50.6621816Z [ERROR]   MySqlWallTest90.test_true:38
2024-01-27T16:08:50.6622404Z [ERROR]   MySqlWallTest91.test:245
2024-01-27T16:08:50.6622762Z [ERROR]   MySqlWallTest91.test_false3:78
2024-01-27T16:08:50.6623138Z [ERROR]   MySqlWallTest91.test_false4:86
2024-01-27T16:08:50.6623786Z [ERROR]   MySqlWallTest91.test_false5:94
2024-01-27T16:08:50.6624223Z [ERROR]   MySqlWallTest91.test_false6:102
2024-01-27T16:08:50.6624755Z [ERROR]   MySqlWallTest_union.testUnion:30
2024-01-27T16:08:50.6625139Z [ERROR]   OracleWallTest.testWall:29
2024-01-27T16:08:50.6625534Z [ERROR]   AntlrMySqlTest.test_for_antlr_examples:39

**Comment by qxo** [comment]
看到了, 还没修改完呢, 因昨codespace网络有问题切换回本地所以临时先提交了

越修改测试用例越有疑问: 如果只是写反了, 怎么可能这么我测试代码其逻辑也是反的....




------------------&nbsp;原始邮件&nbsp;------------------
发件人: "Rossen ***@***.***&gt;; 
发送时间: 2024年1月28日(星期天) 上午8:12
收件人: ***@***.***&gt;; 
抄送: ***@***.***&gt;; ***@***.***&gt;; 
主题: Re: [alibaba/druid] fix: 修复SQL防火墙之SQL条件总真(WhereAlwayTrue)功能无效问题(条件写反了) (PR #5698)





 
fix: 修正测试用户逻辑这个commit有20个测试ERROR：
 2024-01-27T16:08:50.6612964Z [ERROR]   WallUnionTest2.testMySql:34
 2024-01-27T16:08:50.6613602Z [ERROR]   WallUnionTest2.testOracle:43
 2024-01-27T16:08:50.6614202Z [ERROR]   WallUnionTest4.testMySql:34
 2024-01-27T16:08:50.6616489Z [ERROR]   WallUnionTest4.testORACLE:43
 2024-01-27T16:08:50.6617366Z [ERROR]   WallUpdateTest3.testMySql:41
 2024-01-27T16:08:50.6617998Z [ERROR]   WallUpdateTest3.testORACLE:46
 2024-01-27T16:08:50.6618641Z [ERROR]   WallUpdateWhereTest.testMySql:34
 2024-01-27T16:08:50.6619301Z [ERROR]   WallUpdateWhereTest.testORACLE:39
 2024-01-27T16:08:50.6619990Z [ERROR]   WallUpdateWhereTest1.test_check_true:38
 2024-01-27T16:08:50.6620651Z [ERROR]   MySqlWallTest63.test_true:40
 2024-01-27T16:08:50.6621231Z [ERROR]   MySqlWallTest74.test_false:38
 2024-01-27T16:08:50.6621816Z [ERROR]   MySqlWallTest90.test_true:38
 2024-01-27T16:08:50.6622404Z [ERROR]   MySqlWallTest91.test:245
 2024-01-27T16:08:50.6622762Z [ERROR]   MySqlWallTest91.test_false3:78
 2024-01-27T16:08:50.6623138Z [ERROR]   MySqlWallTest91.test_false4:86
 2024-01-27T16:08:50.6623786Z [ERROR]   MySqlWallTest91.test_false5:94
 2024-01-27T16:08:50.6624223Z [ERROR]   MySqlWallTest91.test_false6:102
 2024-01-27T16:08:50.6624755Z [ERROR]   MySqlWallTest_union.testUnion:30
 2024-01-27T16:08:50.6625139Z [ERROR]   OracleWallTest.testWall:29
 2024-01-27T16:08:50.6625534Z [ERROR]   AntlrMySqlTest.test_for_antlr_examples:39

**Comment by zrlw** [comment]
我也不知道为什么会有这么多错误的单元测试，可能这些单元测试案例都是恒真条件检查机制失效后加的。

**Comment by zrlw** [comment]
或许这些错误的单元测试需要将恒真检查选项设为false。

**Comment by CantosSong** [comment]
wall的默认行为，如果从false变成true 会对用到的人有影响，更应该默认是false

**Comment by CantosSong** [comment]
> wall的默认行为，如果从false变成true 会对用到的人有影响，更应该默认是false

当然，updateWhereAlwayTrueCheck 定义没问题，但是原始的默认值是false,修改为true就会对既有项目有影响

**Pull request merged**